### PR TITLE
CUDA 8, OpenCV 2 Supported and Options to disable libjpeg-turbo and nvJPEG

### PR DESCRIPTION
**Pull request by gzygzy9211**
For CUDA 8 Support: 
  a) Disable sm70;
  b) Nppi is a single library
  c) Add Wrapper for type casting, and specialize it for float16 when use CUDA 8
  d) a nppiResize Wrapper for converting CUDA 9 parameters to CUDA 8 version. (Not complete but should work properly for DALI, see the comment for details)

For OpenCV 2 Support:
  use highgui instead of imcodec when using OpenCV 2.x

Options to disable libjpeg-turbo and nvJPEG
  a) add corresponding options in CMakeLists.txt
  b) add definitions in Dependencies.cmake to indicate whether it is available
  c) fallback logic

### CODE REVIEW

**Comment by gzygzy9211** [comment]
Respect to #11 
Sorry I don't know how to add the request to be issue directly

**Comment by JanuszL** [comment]
Thank you for your pull request - that makes you first external developer contributing to Dali!!!
This also reminds me that we should write some guide how to validate code before making such contribution.
In the mean time please run lint:
cd _YOUR_BUILD_DIR_
make lint

File: dali/pipeline/operators/fused/crop_mirror_normalize.cu [hunk 1]
```diff
@@ -55,7 +55,7 @@ __global__ void BatchedCropMirrorNormalizePermuteKernel(
             int in_idx = c + C*mirrored_width + in_step*h;  // HWC, mirrored
             int out_idx = c*H*W + h*W + w;  // CHW
 
-            output_ptr[out_idx] = static_cast<Out>(
+            output_ptr[out_idx] = StaticCastGpu<Out>( // static_cast<Out>(
```
Comment (JanuszL) [inline dali/pipeline/operators/fused/crop_mirror_normalize.cu]: Remove static_cast<Out> instead commenting it out.

File: dali/pipeline/operators/fused/crop_mirror_normalize.cu [hunk 2]
```diff
@@ -68,7 +68,7 @@ __global__ void BatchedCropMirrorNormalizePermuteKernel(
             int in_idx = c + C*w + in_step*h;  // HWC
             int out_idx = c*H*W + h*W + w;  // CHW
 
-            output_ptr[out_idx] = static_cast<Out>(
+            output_ptr[out_idx] = StaticCastGpu<Out>( // static_cast<Out>(
```
Comment (JanuszL) [inline dali/pipeline/operators/fused/crop_mirror_normalize.cu]: Remove static_cast<Out> instead commenting it out.

File: dali/pipeline/operators/fused/crop_mirror_normalize.cu [hunk 3]
```diff
@@ -108,7 +108,7 @@ __global__ void BatchedCropMirrorNormalizePermuteKernel(
         input = (static_cast<float>(input_ptr[in_idx])-mean[c]) * inv_std[c];
       }
 
-      output_ptr[out_idx] = static_cast<Out>(input);
+      output_ptr[out_idx] = StaticCastGpu<Out>(input); // static_cast<Out>(input);
```
Comment (JanuszL) [inline dali/pipeline/operators/fused/crop_mirror_normalize.cu]: Remove static_cast<Out> instead commenting it out.

File: dali/pipeline/operators/fused/normalize_permute.cu [hunk 4]
```diff
@@ -32,7 +32,7 @@ __global__ void BatchedNormalizePermuteKernel(const uint8 *in_batch,
   for (int c = 0; c < C; ++c) {
     for (int h = threadIdx.y; h < H; h += blockDim.y) {
       for (int w = threadIdx.x; w < W; w += blockDim.x) {
-        out[c*H*W + h*W + w] = static_cast<OUT>(
+        out[c*H*W + h*W + w] = StaticCastGpu<OUT>( // static_cast<OUT>(
```
Comment (JanuszL) [inline dali/pipeline/operators/fused/normalize_permute.cu]: Remove static_cast<Out> instead commenting it out.

File: CMakeLists.txt [hunk 5]
```diff
@@ -12,7 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-cmake_minimum_required(VERSION 3.5)
+cmake_minimum_required(VERSION 3.9)
```
Comment (slayton58) [inline CMakeLists.txt]: Why is this necessary? Default on Ubuntu 16.04 is 3.5.1 so we can't go higher than that.

File: dali/image/jpeg.cc [hunk 6]
```diff
@@ -94,9 +96,13 @@ bool get_jpeg_size(const uint8 *data, size_t data_size, int *height, int *width)
 }  // namespace
 
 bool CheckIsJPEG(const uint8 *jpeg, int) {
+// Return false to indicate HostDecoder to use OpenCV fallback
+// see also dali/pipeline/operators/decoder/host_decoder.h:50
+#ifdef DALI_USE_JPEG_TURBO
```
Comment (slayton58) [inline dali/image/jpeg.cc]: Gating this doesn't make sense -- the code doesn't depend on TJPG at all.

File: dali/pipeline/operators/decoder/nvjpeg_decoder.cu [hunk 7]
```diff
@@ -12,6 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#ifdef DALI_USE_NVJPEG
```
Comment (slayton58) [inline dali/pipeline/operators/decoder/nvjpeg_decoder.cu]: Decide whether to build the nvJPEG op via cmake instead of `#ifdef`

File: CMakeLists.txt (hunk 5 above)
Comment (gzygzy9211) [reply CMakeLists.txt]: I have checked the documentation of CMake and put the reasons in the commit comment:

In "dali/CMakeLists.txt":44
CUDA_LINK_LIBRARIES_KEYWORD is available in CMake 3.9 or above

In "dali/pipeline/operators/reader/parser/CMakeLists.txt":22
Parameter "EXPORT_MACRO" for protobuf_generate_cpp is available in CMake 3.9 or above

File: dali/pipeline/operators/fused/crop_mirror_normalize.cu (hunk 1 above)
Comment (gzygzy9211) [reply dali/pipeline/operators/fused/crop_mirror_normalize.cu]: Ok, I will follow this. And I still working on lint. 
Thank you for reviewing

File: CMakeLists.txt (hunk 5 above)
Comment (gzygzy9211) [reply CMakeLists.txt]: I have checked the documentation of CMake and put reasons in the comment of this commit.

In "dali/CMakeLists.txt":44
CUDA_LINK_LIBRARIES_KEYWORD is available in CMake 3.9 or above

In "dali/pipeline/operators/reader/parser/CMakeLists.txt":22
Parameter "EXPORT_MACRO" for protobuf_generate_cpp is available in CMake 3.9 or above

Comment (gzygzy9211) [reply CMakeLists.txt]: For CUDA_LINK_LIBRARIES_KEYWORD, there is a workaround:
set(CUDA_LIBRARIES ${CUDA_LINK_LIBRARIES_KEYWORD} ${CUDA_LIBRARIES})
applying this when cmake is below 3.9

As for EXPORT_MACRO" parameter of protobuf_generate_cpp, I haven't found any solution for CMake below 3.9 for now

File: dali/image/jpeg.cc (hunk 6 above)
Comment (gzygzy9211) [reply dali/image/jpeg.cc]: As explained in comment: 
In dali/pipeline/operators/decoder/host_decoder.h:50, "CheckIsJPEG" is called to decide whether to use TJPG, by returning false decoder will fallback to OpenCV decoding

File: dali/pipeline/operators/decoder/nvjpeg_decoder.cu (hunk 7 above)
Comment (gzygzy9211) [reply dali/pipeline/operators/decoder/nvjpeg_decoder.cu]: Do you mean exclude nvjpeg_decoder.cu in CMakeLists? 
As this source file acts as a wrapper of nvJPEG for DALI,  I choose to "comment out" this whole source to disable nvJPEG. Maybe doing it via cmake will make CMakeLists.txt a bit more complex.
@JanuszL what do you think?

File: dali/image/jpeg.cc (hunk 6 above)
Comment (slayton58) [reply dali/image/jpeg.cc]: That's not the case exactly -- if `CheckIsJPEG` returns true, it'll _try_ to use TJPG for decode, then if it fails it'll fall back and use OCV. You've already made changes to jpeg.cc, it'd make more sense to leave `CheckIsJPEG` as-is (after all, we're not fundamentally changing what is or isn't a JPEG file), and make it call opencv immediately instead of as a fallback.

File: dali/pipeline/operators/decoder/nvjpeg_decoder.cu (hunk 7 above)
Comment (slayton58) [reply dali/pipeline/operators/decoder/nvjpeg_decoder.cu]: Yes. We already use such a method (see LMDB, proto3, etc). It keeps the source files cleaner at the expense of a little more complexity in the CMakeLists.txt

It also means that if you disable nvJPEG then there will be absolutely nothing nvJPEG-related in the build, no files, nothing.

Comment (gzygzy9211) [reply dali/pipeline/operators/decoder/nvjpeg_decoder.cu]: Thank you for your reviewing too.
I will push the commit after finishing lint

Comment (gzygzy9211) [reply dali/pipeline/operators/decoder/nvjpeg_decoder.cu]: Ok, I will do it

File: dali/image/jpeg.cc (hunk 6 above)
Comment (gzygzy9211) [reply dali/image/jpeg.cc]: What I see in host_decoder.h is that return value of "DecodeJPEGHost" is handled by marco "DALI_CALL", and OCV fallback is in another "if" branch, chosen by "CheckIsJPEG" and "CheckIsPNG". 

Could you please explain it in more detail?

Comment (slayton58) [reply dali/image/jpeg.cc]: Ok, so [host_decoder.h](https://github.com/NVIDIA/DALI/blob/master/dali/pipeline/operators/decoder/host_decoder.h#L50-L55) checks on the type of image being decoded. This then calls [DecodeJPEGHost](https://github.com/NVIDIA/DALI/blob/master/dali/image/jpeg.cc#L108) to perform the actual decode.

`DecodeJpegHost` then attempts to decode using TJPG, and if that fails, decodes using OCV. So if TJPG were to be disabled in the build, I would expect `DecodeJPEGHost` to still be called, and then that method to just call OCV. Make sense?

Comment (gzygzy9211) [reply dali/image/jpeg.cc]: Well, I mean currently, in [host_decoder.h](https://github.com/NVIDIA/DALI/blob/master/dali/pipeline/operators/decoder/host_decoder.h#L50-L55), an exception is thrown if ```DecodeJpegHost``` returns a failure, rather than retry decoding via OCV.

So should I change the strategy that using OCV as fallback if Jpeg or Png decoding fails?

Comment (slayton58) [reply dali/image/jpeg.cc]: That's because DecodeJPEGHost (and DecodePNGHost, and any others we may add later) is always expected to succeed on a valid input. It maintains its own fallback path (in this case also OCV) in case TJPG cannot decode the image (or in your case, isn't available).

The code path you're relying on now is for image formats we don't have explicit paths for -- even if TJPG isn't available, we would still want the DecodeJPEGHost path used, and `CheckIsJPEG` to return true if the input is in fact a JPEG.

Comment (gzygzy9211) [reply dali/image/jpeg.cc]: So if I get you right, I should add OCV decoding codes to ```DecodeJpegHost``` in [host_decoder.h](https://github.com/NVIDIA/DALI/blob/master/dali/pipeline/operators/decoder/host_decoder.h) which is enabled when TJPG isn't available. 

I can do that.

Comment (slayton58) [reply dali/image/jpeg.cc]: Yes please :)

**Comment by gzygzy9211** [comment]
Update has been made according to the reviews. 
:）

File: dali/pipeline/operators/reader/parser/CMakeLists.txt [hunk 8]
```diff
@@ -29,7 +35,12 @@ SET_TARGET_PROPERTIES(CAFFE2_PROTO PROPERTIES COMPILE_FLAGS "-include dali/api_h
 
 # TF
 if (BUILD_PROTO3)
-  protobuf_generate_cpp(TF_PROTO_SRCS TF_PROTO_HEADERS EXPORT_MACRO DLL_PUBLIC proto/example.proto proto/feature.proto)
+  if(${CMAKE_VERSION} VERSION_LESS "3.9.0")
```
Comment (JanuszL) [inline dali/pipeline/operators/reader/parser/CMakeLists.txt]: This way you expose symbols we want to hide. I will address this by providing own protobuf_generate_cpp version shipped with Dali.

Comment (JanuszL) [inline dali/pipeline/operators/reader/parser/CMakeLists.txt]: This way you expose all protobuf symbols we want to hide. I will address this by providing custom protobuf_generate_cpp shipped with Dali.

Comment (gzygzy9211) [reply dali/pipeline/operators/reader/parser/CMakeLists.txt]: so is there any thing I need to do?

Comment (gzygzy9211) [reply dali/pipeline/operators/reader/parser/CMakeLists.txt]: If the target is to hide all symbols in ```*.pb.cc```, the universal way is adding a static library for every ```.proto``` file, and DALI will deal with these library in the way dealing with CUDA static library. 
Or only the implementation of the protobuf message is expected to be hidden?

**Comment by gzygzy9211** [comment]
@JanuszL I just copy the ```PROTOBUF_GENERATE_CPP``` function (renamed with "DALI_" prefix) from CMake 3.9 and have checked it working

File: dali/image/jpeg.cc [hunk 9]
```diff
@@ -118,10 +131,6 @@ DALIError_t DecodeJPEGHost(const uint8 *jpeg, int size,
   } else {
     DALI_RETURN_ERROR("Unsupported image type.");
   }
-  int h, w;
-  int c = (type == DALI_GRAY) ? 1 : 3;
-
-  DALI_CALL(GetJPEGImageDims(jpeg, size, &h, &w));
 
 #ifndef NDEBUG
```
Comment (JanuszL) [inline dali/image/jpeg.cc]: Please move this debug code to line 116 - between GetJPEGImageDims and image->Resize

File: cmake/ProtobufGenCpp.cmake [hunk 10]
```diff
@@ -0,0 +1,97 @@
+# Shipped from CMake 3.9.6, with small modification
```
Comment (JanuszL) [inline cmake/ProtobufGenCpp.cmake]: Please update Acknowledgements.txt accordingly (as we takes pieces from CMake).

Comment (gzygzy9211) [reply cmake/ProtobufGenCpp.cmake]: To make sure, if it is ok to add the Copyright content of CMake to ACK

File: CMakeLists.txt [hunk 11]
```diff
@@ -62,9 +65,14 @@ set(CUDA_NVCC_FLAGS "${CUDA_NVCC_FLAGS} \
   -gencode arch=compute_50,code=sm_50 \
   -gencode arch=compute_52,code=sm_52 \
   -gencode arch=compute_60,code=sm_60 \
-  -gencode arch=compute_61,code=sm_61 \
-  -gencode arch=compute_70,code=sm_70 \
-  -gencode arch=compute_70,code=compute_70")
+  -gencode arch=compute_61,code=sm_61 ")
+#  -gencode arch=compute_70,code=sm_70 \
```
Comment (JanuszL) [inline CMakeLists.txt]: Remove this commented code.

File: cmake/ProtobufGenCpp.cmake (hunk 10 above)
Comment (gzygzy9211) [reply cmake/ProtobufGenCpp.cmake]: Anything I need to do ?

**Comment by JanuszL** [comment]
Hi,
Sorry for keep you waiting. We need some time to prepare guidelines and rules recording contribution from outside of Nvidia.
You can check guide in https://github.com/NVIDIA/DALI/pull/20.
When it is merged we would ask you to sign-off you commit so you confirm you have right to do such contribution

**Comment by gzygzy9211** [comment]
false operation, I have some issues during signing-off
